confd/pkg/backends/calico/client.go

```diff
@@ -825,15 +832,48 @@ func (c *client) onUpdates(updates []api.Update, needUpdatePeersV1 bool) {
 			for _, kvp := range kvps {
 				log.Debugf("KVP: %#v", kvp)
 				if kvp.Value == nil {
+					// Remove node's IPs from our node IP cache
```

Comment:
I think you don't need to check if the entry is in the map or not - you can just do this:

```go
delete(c.nodeIPs, nodeIPv4)
delete(c.nodeIPs, nodeIPv6)
```

You might also want to introduce a helper struct to manage this. 

e.g., `MarkNodeIP(ip string)` and `ClearNodeIP(ip string)`

I only bring this up because confd runs across multiple goroutines so there is some risk of concurrent access issues with maps like this, and helper structs often make it easier to ensure access is done safely.

Reply:
From my understanding, all of the reads/updates for `nodeIPs` should fall under the `cacheLock` whenever any updates are required so modifying the node IP cache should be safe (but only because of where it is called from).